pkg/bloomgateway/bloomgateway.go

```diff
@@ -339,67 +341,45 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
 
 	tasks := make([]Task, 0, len(seriesWithBloomsPerDay))
 	for _, seriesWithBounds := range seriesWithBloomsPerDay {
-		task, err := NewTask(tenantID, seriesWithBounds, req.Filters)
+		task, err := NewTask(ctx, tenantID, seriesWithBounds, req.Filters)
 		if err != nil {
 			return nil, err
 		}
 		tasks = append(tasks, task)
-		expectedResponses += len(seriesWithBounds.series)
+		numSeries += len(seriesWithBounds.series)
 	}
 
 	g.activeUsers.UpdateUserTimestamp(tenantID, time.Now())
 
-	errCh := make(chan error, 1)
-	resCh := make(chan v1.Output, 1)
+	tasksCh := make(chan Task)
 
 	for _, task := range tasks {
-		level.Info(g.logger).Log("msg", "enqueue task", "task", task.ID, "day", task.day, "series", len(task.series))
+		task := task
+		level.Info(logger).Log("msg", "enqueue task", "task", task.ID, "day", task.day, "series", len(task.series))
 		g.queue.Enqueue(tenantID, []string{}, task, func() {
 			// When enqueuing, we also add the task to the pending tasks
 			g.pendingTasks.Add(task.ID, task)
 		})
-
-		// Forward responses or error to the main channels
-		// TODO(chaudum): Refactor to make tasks cancelable
-		go func(t Task) {
-			for {
-				select {
-				case <-ctx.Done():
-					return
-				case err := <-t.ErrCh:
-					if ctx.Err() != nil {
-						level.Warn(g.logger).Log("msg", "received err from channel, but context is already done", "err", ctx.Err())
-						return
-					}
-					errCh <- err
-				case res := <-t.ResCh:
-					level.Debug(g.logger).Log("msg", "got partial result", "task", t.ID, "tenant", tenantID, "fp_int", uint64(res.Fp), "fp_hex", res.Fp, "chunks_to_remove", res.Removals.Len())
-					if ctx.Err() != nil {
-						level.Warn(g.logger).Log("msg", "received res from channel, but context is already done", "err", ctx.Err())
-						return
-					}
-					resCh <- res
-				}
-			}
-		}(task)
+		go consumeTask(ctx, task, tasksCh, logger)
 	}
 
-	responses := responsesPool.Get(expectedResponses)
+	responses := responsesPool.Get(numSeries)
 	defer responsesPool.Put(responses)
+	remaining := len(tasks)
 
 outer:
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, errors.Wrap(ctx.Err(), "waiting for results")
-		case err := <-errCh:
-			return nil, errors.Wrap(err, "waiting for results")
-		case res := <-resCh:
-			responses = append(responses, res)
-			// log line is helpful for debugging tests
-			level.Debug(g.logger).Log("msg", "got partial result", "progress", fmt.Sprintf("%d/%d", len(responses), expectedResponses))
-			// wait for all parts of the full response
-			if len(responses) == expectedResponses {
+			return nil, errors.Wrap(ctx.Err(), "request failed")
+		case task := <-tasksCh:
+			level.Info(logger).Log("msg", "task done", "task", task.ID, "err", task.Err())
+			if task.Err() != nil {
```

Comment:
in this case, the first failing task will break the execution and we won't read from `tasksCh` anymore and it will mean that all goroutines started here `go consumeTask(ctx, task, tasksCh, logger)` , will be blocked.... 
unless the channel has enough capacity to keep all the responses, regardless they will be read or not...

Reply:
I tried to reproduce this with[ this test case](https://github.com/grafana/loki/pull/11792/files?diff=unified&w=0#diff-2783be6588fd384b385d0daa07cac14a9d48891dbb96b3eca7c11a57f8666b94R187), but it does not fail.

Reply:
I changed it and added a comment.